src/Components/Blazor/WebAssembly.Authentication/src/AuthenticationManager.cs

```diff
@@ -0,0 +1,446 @@
+                switch (result.Status)
+                {
+                    case RemoteAuthenticationStatus.Redirect:
+                        break;
+                    case RemoteAuthenticationStatus.Success:
+                        await NavigateToReturnUrl(returnUrl);
+                        break;
+                    case RemoteAuthenticationStatus.OperationCompleted:
+                        break;
+                    case RemoteAuthenticationStatus.Failure:
+                        var uri = Navigation.ToAbsoluteUri($"{ApplicationPaths.LogoutFailedPath}?message={result.ErrorMessage}").ToString();
+                        Navigation.NavigateTo(uri);
+                        break;
+                    default:
+                        throw new InvalidOperationException($"Invalid authentication result status '{result.Status ?? "(null)"}'.");
+                }
+            }
+        }
+
+        private async Task ProcessLogoutCallback()
+        {
+            var result = await AuthenticationService.CompleteSignOutAsync(new RemoteAuthenticationContext<TAuthenticationState> { Url = Navigation.Uri });
+            switch (result.Status)
+            {
+                case RemoteAuthenticationStatus.Redirect:
+                    // There should not be any redirects as the only time completeAuthentication finishes
+                    // is when we are doing a redirect sign in flow.
+                    throw new InvalidOperationException("Should not redirect.");
+                case RemoteAuthenticationStatus.Success:
+                    await NavigateToReturnUrl(GetReturnUrl(result.State, Navigation.ToAbsoluteUri(ApplicationPaths.LogoutSucceededPath).ToString()));
+                    break;
+                case RemoteAuthenticationStatus.OperationCompleted:
+                    break;
+                case RemoteAuthenticationStatus.Failure:
+                    var uri = Navigation.ToAbsoluteUri($"{ApplicationPaths.LogoutFailedPath}?message={result.ErrorMessage}").ToString();
+                    Navigation.NavigateTo(uri);
+                    break;
+                default:
+                    throw new InvalidOperationException($"Invalid authentication result status '{result.Status ?? "(null)"}'.");
+            }
+        }
+
+        private string GetReturnUrl(TAuthenticationState state, string defaultReturnUrl = null)
+        {
+            var fromQuery = GetParameter("returnUrl");
+            if (!string.IsNullOrWhiteSpace(fromQuery) && !fromQuery.StartsWith(Navigation.BaseUri))
+            {
+                // This is an extra check to prevent open redirects.
+                throw new InvalidOperationException("Invalid return url. The return url needs to have the same origin as the current page.");
+            }
+
+            return state?.ReturnUrl ?? fromQuery ?? defaultReturnUrl ?? Navigation.BaseUri;
```

Comment:
Is it really legal for `state` to be null here? In other places this class assumes it's nonnull.

Reply:
The link isn't working for me but I'll believe you!